lite/inquiring_certifier.go

```diff
@@ -1,163 +1,216 @@
+	if err := nfc.ValidateFull(ic.chainID); err != nil {
+		return err
 	}
-	return err
+	// Trust it.
+	return ic.trusted.SaveFullCommit(nfc)
 }
 
-func (ic *InquiringCertifier) useClosestTrust(h int64) error {
-	closest, err := ic.trusted.GetByHeight(h)
+// verifyAndSave will verify if this is a valid source full commit given the
+// best match trusted full commit, and if good, persist to ic.trusted.
+// Returns ErrTooMuchChange when >2/3 of tfc did not sign sfc.
+// Panics if tfc.Height() >= sfc.Height().
+func (ic *InquiringCertifier) verifyAndSave(tfc, sfc FullCommit) error {
+	if tfc.Height() >= sfc.Height() {
+		panic("should not happen")
+	}
+	err := tfc.NextValidators.VerifyFutureCommit(
+		sfc.Validators,
+		ic.chainID, sfc.SignedHeader.Commit.BlockID,
+		sfc.SignedHeader.Height, sfc.SignedHeader.Commit,
+	)
 	if err != nil {
 		return err
 	}
 
-	// if the best seed is not the one we currently use,
-	// let's just reset the dynamic validator
-	if closest.Height() != ic.LastHeight() {
-		ic.cert = NewDynamicCertifier(ic.ChainID(), closest.Validators, closest.Height())
-	}
-	return nil
+	return ic.trusted.SaveFullCommit(sfc)
 }
 
-// updateToHash gets the validator hash we want to update to
-// if IsTooMuchChangeErr, we try to find a path by binary search over height
-func (ic *InquiringCertifier) updateToHash(vhash []byte) error {
-	// try to get the match, and update
-	fc, err := ic.Source.GetByHash(vhash)
+// updateToHeight will use divide-and-conquer to find a path to h.
+// Returns nil error iff we successfully verify and persist a full commit
+// for height h, using repeated applications of bisection if necessary.
+//
+// Returns ErrCommitNotFound if source provider doesn't have the commit for h.
+func (ic *InquiringCertifier) updateToHeight(h int64) (FullCommit, error) {
+
+	// Fetch latest full commit from source.
+	sfc, err := ic.source.LatestFullCommit(ic.chainID, h, h)
 	if err != nil {
-		return err
+		return FullCommit{}, err
 	}
-	err = ic.cert.Update(fc)
-	// handle IsTooMuchChangeErr by using divide and conquer
-	if liteErr.IsTooMuchChangeErr(err) {
-		err = ic.updateToHeight(fc.Height())
-	}
-	return err
-}
 
-// updateToHeight will use divide-and-conquer to find a path to h
-func (ic *InquiringCertifier) updateToHeight(h int64) error {
-	// try to update to this height (with checks)
-	fc, err := ic.Source.GetByHeight(h)
-	if err != nil {
-		return err
+	// Validate the full commit.  This checks the cryptographic
+	// signatures of Commit against Validators.
+	if err := sfc.ValidateFull(ic.chainID); err != nil {
+		return FullCommit{}, err
 	}
-	start, end := ic.LastHeight(), fc.Height()
-	if end <= start {
-		return liteErr.ErrNoPathFound()
+
+	// If sfc.Height() != h, we can't do it.
+	if sfc.Height() != h {
+		return FullCommit{}, lerr.ErrCommitNotFound()
 	}
-	err = ic.Update(fc)
 
-	// we can handle IsTooMuchChangeErr specially
-	if !liteErr.IsTooMuchChangeErr(err) {
-		return err
+FOR_LOOP:
```

Comment:
This loop can probably be rewritten to reduce number of LastCommit fetches.

Reply:
Saving optimizations for later.  This one isn't too bad, still has the O() property we want, though the whole scheme could improve w/ Tendermint streaming memoized valset changes to the lite client (rather than any binary search).